OSDOCS-9171#Add SMB CIFS CSI Driver Operator

lpettyjo opened the pull request:
<!--- PR title format: [GH#<gh-issue-id>][BZ#<bz-issue-id>][OCPBUGS#<jira-issue-id>][OSDOCS#<jira-issue-id>]: <short-description-of-the-pr> --->

<!--- If your changes apply to the latest release and/or in-development version of OpenShift, open your PR against the `main` branch.

* For more details about the information requested in this template, see:
  https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/create_or_edit_content.adoc#submit-PR --->

Version(s): 4.16+
<!--- Specify the version or versions of OpenShift your PR applies to. -->

Issue: https://issues.redhat.com/browse/OSDOCS-9171
<!--- Add a link to the Bugzilla, Jira, or GitHub issue, if applicable. --->

Link to docs preview:
https://75920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi-smb-cifs
https://75920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/storage/understanding-persistent-storage
https://75920--ocpdocs-pr.netlify.app/openshift-enterprise/latest/storage/container_storage_interface/persistent-storage-csi#csi-drivers-supported_persistent-storage-csi
<!--- Add direct link(s) to the exact page(s) with updated content from the preview build. --->

QE review:
- [ ] QE has approved this change.
<!--- QE approval is required to merge a PR except for changes that do not impact the meaning of the docs. --->

Additional information:
<!--- Optional: Include additional context or expand the description here.--->

<!--- After you open your PR, ask for review from the OpenShift docs team:
  For community authors: Tag @openshift/team-documentation in a GitHub comment.--->

PTAL: @gcharot @mpatlasov @Phaow

bscott-rh commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,38 @@
+mountOptions:
+  - mfsymlinks
+--
+
```
You may want to add a second step here showing how to apply the yaml file, such as step 3 in [creating a compute machine set](https://docs.openshift.com/container-platform/4.15/machine_management/creating_machinesets/creating-machineset-azure.html#machineset-creating_creating-machineset-azure).

bscott-rh reviewed (commented):
Left 3 comments, otherwise LGTM.

Phaow commented on modules/persistent-storage-csi-olm-operator-install.adoc:
```diff
@@ -2,12 +2,29 @@
 
-The link:https://github.com/openshift/aws-efs-csi-driver-operator[AWS EFS CSI Driver Operator] (a Red Hat operator) is not installed in {product-title} by default. Use the following procedure to install and configure the {FeatureName} CSI Driver Operator in your cluster.
+The {FeatureName} CSI Driver Operator (a Red Hat operator) is not installed in {product-title} by default. Use the following procedure to install and configure the {FeatureName} CSI Driver Operator in your cluster.
+
```
After Installing the SMB CIFS CSI Driver Operator we need to do the similar thing to create the smb clustercsidriver to trigger the csidriver install.(ref: https://docs.openshift.com/container-platform/4.15/storage/container_storage_interface/persistent-storage-csi-aws-efs.html#persistent-storage-csi-efs-driver-install_persistent-storage-csi-aws-efs)
```console
apiVersion: operator.openshift.io/v1
kind: ClusterCSIDriver
metadata:
    name: smb.csi.k8s.io
spec:
  managementState: Managed

```

Phaow commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,34 @@
+
+Familiarity with xref:../../storage/understanding-persistent-storage.adoc#understanding-persistent-storage[persistent storage] and xref:../../storage/container_storage_interface/persistent-storage-csi.adoc#persistent-storage-csi[configuring CSI volumes] is recommended when working with a CSI Operator and driver.
+
+After installing the SMB CIFS CSI Driver Operator, {product-title} installs the SMB CIFS CSI Operator and the SMB CIFS CSI driver by default in the `openshift-cluster-csi-drivers` namespace. This allows the SMB CIFS CSI Driver Operator to create CSI-provisioned PVs that mount to SMB CIFS assets.
```
After Installing the SMB CIFS CSI Driver Operator we need to do the similar thing to create the smb clustercsidriver to trigger the csidriver install.(ref: https://docs.openshift.com/container-platform/4.15/storage/container_storage_interface/persistent-storage-csi-aws-efs.html#persistent-storage-csi-efs-driver-install_persistent-storage-csi-aws-efs)
```console
apiVersion: operator.openshift.io/v1
kind: ClusterCSIDriver
metadata:
    name: smb.csi.k8s.io
spec:
  managementState: Managed

```

mpatlasov replied:
For example, for GCP Filestore, this comes as last step in "Installing the GCP Filestore CSI Driver Operator" immediately before "Creating a storage class for GCP Filestore Storage":

> 4. Install the GCP Filestore CSI Driver:
> a. Click administration → CustomResourceDefinitions → ClusterCSIDriver.
> b. On the Instances tab, click Create ClusterCSIDriver.

(https://docs.openshift.com/container-platform/4.15/storage/container_storage_interface/persistent-storage-csi-google-cloud-file.html)

We do need similar wording for SBM, imho.

Phaow commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,34 @@
+
+After installing the SMB CIFS CSI Driver Operator, {product-title} installs the SMB CIFS CSI Operator and the SMB CIFS CSI driver by default in the `openshift-cluster-csi-drivers` namespace. This allows the SMB CIFS CSI Driver Operator to create CSI-provisioned PVs that mount to SMB CIFS assets.
+
+* The _SMB CIFS CSI Driver Operator_, after being installed, does not create a storage class by default to use to create persistent volume claims (PVCs). However, you can manually create the SMB CIFS `StorageClass`. The SMB CIFS CSI Driver Operator supports dynamic volume provisioning by allowing storage volumes to be created on-demand.
```
We also need some extra declarion for the limiation and known issues for smb csi driver. 

- Fips enable cluster doesn’t support
https://access.redhat.com/solutions/256053
- Http proxy config does not support
- `onDelete: archive` does not suggested, since still have some bugs https://issues.redhat.com/browse/OCPBUGS-33474

mpatlasov replied:
As for `onDelete: archive`, we'd like to discourage using archive mode because upstream has important fixes which we have not included (for reason) into OCP for now.

mpatlasov replied:
Also, it would be nice to make "you can manually create the SMB CIFS StorageClass" a hyperlink to a section about StorageClass (as in https://docs.openshift.com/container-platform/4.15/storage/container_storage_interface/persistent-storage-csi-google-cloud-file.html)

lpettyjo replied:
@Phaow @mpatlasov Are the two support issues listed above likely to be resolved? If so, I'd recommend putting them in the RNs as bugs. However, if they are likely to persist, then I could add them as not supported items in the driver topic. 

I've added the xref.

mpatlasov replied:
`onDelete: archive` is going to be resolved soon. Not sure about FIPS and http proxy...

Phaow replied:
For FIPS and http proxy, current SMB protocal does not support yet not sure it will support in the future.

lpettyjo replied:
@Phaow @mpatlasov Thanks! I'll document FIPS & HTTP proxy limitation, and monitor the `oneDelete:archive `situation to see if we need to add that to the RNs.

mpatlasov commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,34 @@
+
+Familiarity with xref:../../storage/understanding-persistent-storage.adoc#understanding-persistent-storage[persistent storage] and xref:../../storage/container_storage_interface/persistent-storage-csi.adoc#persistent-storage-csi[configuring CSI volumes] is recommended when working with a CSI Operator and driver.
+
+After installing the SMB CIFS CSI Driver Operator, {product-title} installs the SMB CIFS CSI Operator and the SMB CIFS CSI driver by default in the `openshift-cluster-csi-drivers` namespace. This allows the SMB CIFS CSI Driver Operator to create CSI-provisioned PVs that mount to SMB CIFS assets.
```
> After installing the SMB CIFS CSI Driver Operator, OpenShift Container Platform installs the SMB CIFS CSI Operator and the SMB CIFS CSI driver by default in the openshift-cluster-csi-drivers namespace. This allows the SMB CIFS CSI Driver Operator to create CSI-provisioned PVs that mount to SMB CIFS assets.

This sounds rather confusing to me. AFAIU, we have two concepts: "CSI Driver" and "CSI Driver Operator". It's hard to guess what's the difference between "SMB CIFS CSI Operator" and "SMB CIFS CSI Driver Operator". Also, it is CSI Driver (not CSI Driver Operator) who creates CSI-provisioned PVs that mount to SMB CIFS assets. I'd suggest the following text:

> After installing the SMB CIFS CSI Driver Operator, OpenShift Container Platform installs corresponding Pods for the operator and driver in the openshift-cluster-csi-drivers namespace by default. This allows the SMB CIFS CSI Driver to create CSI-provisioned PVs that mount to SMB CIFS assets.

gcharot replied:
yes +1

mpatlasov commented on modules/persistent-storage-csi-olm-operator-install.adoc:
```diff
@@ -24,24 +41,42 @@ To install the {FeatureName} CSI Driver Operator from the web console:
 .. Locate the {FeatureName} CSI Operator by typing *{FeatureName} CSI* in the filter box.
```
We probably need to rename `{FeatureName}` from "SMB CIFS CSI" to "CIFS/SMB CSI". I.e. if you try to type "SMB CIFS CSI" in the console, it finds nothing.

gcharot replied:
yes agree let's be accurate with the name

mpatlasov commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,38 @@
+  name: samba
+provisioner: smb.csi.k8s.io
+parameters:
+  source: //samba-server.samba-server.svc.cluster.local/share
```
I'd suggest some explanation for this line:

> This assumes that Samba server is installed somewhere and reachable from the cluster by the hostname `samba-server.samba-server.svc.cluster.local`. And the server is configured to have the path `/share` among exported shares.

gcharot replied:
yes +1 let's be exhaustive on the pre-requisites

mpatlasov commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,38 @@
+  csi.storage.k8s.io/provisioner-secret-name: smbcreds
+  csi.storage.k8s.io/provisioner-secret-namespace: samba-server
+  csi.storage.k8s.io/node-stage-secret-name: smbcreds
+  csi.storage.k8s.io/node-stage-secret-namespace: samba-server
```
These four lines assume that the user put username/password for access to Samba sever to a Secret with name `smbcreds` in the namespace `samba-server`:

```
apiVersion: v1
kind: Secret
metadata:
  name: smbcreds
  namespace: samba-server
stringData:
  username: <put_username_here>
  password: <put_password_here>
```

I think we need such an explanation about `smbcreds` and `samba-server` to make the doc useful.

gcharot replied:
+1 we need a more generic snippet and detailed explanations below

mpatlasov commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,38 @@
+After installing the Operator, you should create a storage class for dynamic provisioning of Server Message Block (SMB) protocol Common Internet File System (CIFS) dialect volumes.
+
+.Prerequisites
+* You are logged in to the running {product-title} cluster.
```
Let's two more prerequisites:

> * You have SMB server installed and know its:
>  - hostname
>  - share name
>  - username/password
> * You created a Secret called `smbcreds` in the namespace `samba-server` and put username and password as `stringData` there.

lpettyjo replied:
I added the 1st prerequiste. I have the 2nd one as part of the procedure.

Phaow commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,38 @@
+reclaimPolicy: Delete
+volumeBindingMode: Immediate
+mountOptions:
+  - mfsymlinks
```
The suggested moutOptions should be as follows
```suggestion
  - dir_mode=0777
  - file_mode=0777
  - noperm
  - mfsymlinks
  - cache=strict
  - noserverino
```
The suggested moutOptions should be

Phaow commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,34 @@
+
+{product-title} is capable of provisioning persistent volumes (PVs) using the Container Storage Interface (CSI) driver for Server Message Block (SMB) protocol Common Internet File System (CIFS) dialect.
+
+:FeatureName: SMB CIFS CSI Driver Operator
```
```suggestion
:FeatureName: CIFS/SMB CSI Driver Operator
```
Maybe better to keep the same name with the display name https://github.com/openshift/csi-operator/blob/79ca0ac2df11f8e047e3eae4e08f13bf745814f1/Dockerfile.samba#L10 .

gcharot commented on modules/storage-persistent-storage-block-volume.adoc:
```diff
@@ -48,6 +48,7 @@ endif::openshift-dedicated,openshift-rosa[]
 ifndef::openshift-dedicated,openshift-rosa[]
 |NFS | | |
 |{rh-storage-first} | ✅ | ✅ | ✅
+|SMB CIFS  |  |  | 
```
we need to fill up the the blanks here, dynamic & static have been tested by QE
cc @mpatlasov @Phaow

Phaow replied:
Hmm, this section is raw block volume support matrix, so from my understanding it should be all not supported. It seems this section also does not unify the unsupport symbol. I saw in `Access modes
` support we use `-` for unsupported, but this raw block just use empty to indicate unsupported. We'd better unify the doc fomat.
```suggestion
|SMB CIFS  | - | - | -
```

lpettyjo replied:
I've standardized on checkmarks and blanks. Also, I've filled in the blanks for the block vol table.

gcharot commented:
@lpettyjo i believe we also need to add it to the "type of PVs" section of "Understanding persistent storage)"

gcharot commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,35 @@
+
+Familiarity with xref:../../storage/understanding-persistent-storage.adoc#understanding-persistent-storage[persistent storage] and xref:../../storage/container_storage_interface/persistent-storage-csi.adoc#persistent-storage-csi[configuring CSI volumes] is recommended when working with a CSI Operator and driver.
+
+After installing the SMB CIFS CSI Driver Operator, {product-title} installs the SMB CIFS CSI Operator and the SMB CIFS CSI driver by default in the `openshift-cluster-csi-drivers` namespace. This allows the SMB CIFS CSI Driver Operator to create CSI-provisioned PVs that mount to SMB CIFS assets.
```
why do we say SMB CIFS assets? shouldn't we use "shares" instead?

gcharot commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,43 @@
+[id="persistent-storage-csi-smb-cifs-create-sc_{context}"]
+= Creating a storage class for SMB CIFS
+
+After installing the Operator, you should create a storage class for dynamic provisioning of Server Message Block (SMB) protocol Common Internet File System (CIFS) dialect volumes.
```
I would like to see a second section that explains how to manually create a PV that points to an existing share.
This will be a common use case where OCP consume and existing CIFS share that already contains data.
https://github.com/kubernetes-csi/csi-driver-smb/blob/master/deploy/example/e2e_usage.md#option2-pvpvc-usage

lpettyjo commented:
@gcharot @mpatlasov @Phaow Comments incorporated, plus new Section "Creating PVs/PVCs with SMB shares". PTAL. Thanks!

Phaow commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,191 @@
+      source: //<hostname>/<shares> <4>
+    nodeStageSecretRef:
+      name: <secret-name> <5>
+      namespace: default
```
```suggestion
      namespace: <secret-namespace> <6>
```
Maybe better, the namespace also could be not the default ns.

lpettyjo replied:
Done.

lpettyjo commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,191 @@
+        "kubernetes.io/os": linux
+      containers:
+        - name: <deployment-name> <1>
+          image: mcr.microsoft.com/oss/nginx/nginx:1.19.5
```
Fixed.

skopacz1 commented on modules/persistent-storage-csi-drivers-supported.adoc:
```diff
@@ -28,26 +28,27 @@ and which CSI features they support, such as volume snapshots and resize.
-|Microsoft Azure Stack Hub | ✅ | ✅ | ✅| -
+|Microsoft Azure Disk | ✅ | ✅ | ✅|  
+|Microsoft Azure Stack Hub | ✅ | ✅ | ✅|  
 |Microsoft Azure File | - | - | ✅| ✅
```
Just an FYI that if you're doing a sweep of unnecessary hyphens then this line still has some

skopacz1 commented on modules/persistent-storage-csi-olm-operator-install.adoc:
```diff
@@ -2,12 +2,29 @@
 = Installing the {FeatureName} CSI Driver Operator
 
-The link:https://github.com/openshift/aws-efs-csi-driver-operator[AWS EFS CSI Driver Operator] (a Red Hat operator) is not installed in {product-title} by default. Use the following procedure to install and configure the {FeatureName} CSI Driver Operator in your cluster.
+The {FeatureName} CSI Driver Operator (a Red Hat operator) is not installed in {product-title} by default. Use the following procedure to install and configure the {FeatureName} CSI Driver Operator in your cluster.
```
Operator needs to be capitalized here:
```suggestion
The {FeatureName} CSI Driver Operator (a Red Hat Operator) is not installed in {product-title} by default. Use the following procedure to install and configure the {FeatureName} CSI Driver Operator in your cluster.
```

skopacz1 replied:
Also per the [SSG](https://redhat-documentation.github.io/supplementary-style-guide/#non-breaking-spaces) there should be a non breaking space in between Red and Hat:
```suggestion
The {FeatureName} CSI Driver Operator (a Red{nbsp}Hat operator) is not installed in {product-title} by default. Use the following procedure to install and configure the {FeatureName} CSI Driver Operator in your cluster.
```

skopacz1 commented on modules/persistent-storage-csi-olm-operator-install.adoc:
```diff
@@ -24,24 +41,42 @@ To install the {FeatureName} CSI Driver Operator from the web console:
+ifdef::restricted[]
 ifdef::openshift-rosa,openshift-enterprise[]
 * If you are using {FeatureName} with AWS Secure Token Service (STS), in the *role ARN* field, enter the ARN role copied from the last step of the _Obtaining a role Amazon Resource Name for Security Token Service_ procedure.
 endif::[]
```
I think it'll be fine without it and it's up to you, but since you have multiple ifdefs right here, it might be good to clarify which attributes close this one out:
```suggestion
endif::openshift-rosa,openshift-enterprise[]
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+
+.Procedure
+
+. Create a PV:
++
+Run the following command with the following example YAML file:
```
I personally feel it reads a bit weird to have two sentences like this ending with a colon in one step, and also since the command has a user replaced value, I feel like you don't have to mention "example". I'd suggest the following rewording:
```suggestion
. Create a PV by running the following command:
```
I'm not sure whether this maintains the technical accuracy of telling users to use the sample information below, though.

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
++
+Run the following command with the following example YAML file:
++
+[source,cli]
```
Per the [doc guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#source-tags-for-terminal-commands-and-output) I think the source type should be terminal:
```suggestion
[source,terminal]
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
++
+[source,cli]
+----
+oc create -f <pv-file-name>.yaml <1>
```
Also per [the guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference), I think user replaced values should have underscores in between words:
```suggestion
oc create -f <pv_file_name>.yaml <1>
```

skopacz1 replied:
This command is missing the $ prompt:
```suggestion
$ oc create -f <pv-file-name>.yaml <1>
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+metadata:
+  annotations:
+    pv.kubernetes.io/provisioned-by: smb.csi.k8s.io
+  name: <pv-name> <1>
+spec:
+  capacity:
+    storage: 100Gi
+  accessModes:
+    - ReadWriteMany
+  persistentVolumeReclaimPolicy: Retain
+  storageClassName: <sc-name> <2>
+  mountOptions:
+    - dir_mode=0777
+    - file_mode=0777
+  csi:
+    driver: smb.csi.k8s.io
+    volumeHandle: smb-server.default.svc.cluster.local/share## <3>
+    volumeAttributes:
+      source: //<hostname>/<shares> <4>
+    nodeStageSecretRef:
+      name: <secret-name> <5>
```
Same comment about using underscores:
```suggestion
  name: <pv_name> <1>
spec:
  capacity:
    storage: 100Gi
  accessModes:
    - ReadWriteMany
  persistentVolumeReclaimPolicy: Retain
  storageClassName: <sc_name> <2>
  mountOptions:
    - dir_mode=0777
    - file_mode=0777
  csi:
    driver: smb.csi.k8s.io
    volumeHandle: smb-server.default.svc.cluster.local/share## <3>
    volumeAttributes:
      source: //<hostname>/<shares> <4>
    nodeStageSecretRef:
      name: <secret_name> <5>
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+----
+<1> The name of the PV.
+<2> The name of the CIFS/SMB storage class that you created earlier (see _Prerequisites_).
+<3> `volumeHandle` format: {smb-server-address}#{sub-dir-name}#{share-name}. Ensure that this value is unique for every share in the cluster.
```
I feel like even if the format does not exactly contain text you'd find in a file, it maybe should be put in backticks since it kind of represents text that would end up in the YAML. What do you think?
```suggestion
<3> `volumeHandle` format: `{smb-server-address}#{sub-dir-name}#{share-name}`. Ensure that this value is unique for every share in the cluster.
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+    - file_mode=0777
+  csi:
+    driver: smb.csi.k8s.io
+    volumeHandle: smb-server.default.svc.cluster.local/share## <3>
```
Just an FYI that I believe the second # is not showing up in the rendered output, I believe it's related to how the # character is used before callouts to ensure that the formatting after the callout still has correct coloring. I think there's some sort of escape character you could put in here to make sure that doesn't happen, maybe a `/`?

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+<1> The name of the PV.
+<2> The name of the CIFS/SMB storage class that you created earlier (see _Prerequisites_).
+<3> `volumeHandle` format: {smb-server-address}#{sub-dir-name}#{share-name}. Ensure that this value is unique for every share in the cluster.
+<4> The Samba server must be installed somewhere and reachable from the cluster with <hostname> being the hostname for the Samba server and <shares> the path the server is configured to have among the exported shares.
```
I think these values should have backticks since they directly reference what's in the YAML:
```suggestion
<4> The Samba server must be installed somewhere and reachable from the cluster with `<hostname>` being the hostname for the Samba server and `<shares>` the path the server is configured to have among the exported shares.
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+
+.. Create a PVC by running the following command with the following example YAML file:
++
+[source, cli]
```
Same comment about terminal:
```suggestion
[source, terminal]
```

I see `cli` used in other commands in this PR, so rather than me repeating this feedback, please just go through the PR to make sure all of them say `terminal` instead. Thanks!

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
++
+[source, cli]
+----
+oc create -f <pv-file-name>.yaml <1>
```
Same comment about $:
```suggestion
$ oc create -f <pv-file-name>.yaml <1>
```

Please also go through the commands to make sure they all have $ in front of them

skopacz1 replied:
Same comment about underscores in between the words of user replaced values:
```suggestion
oc create -f <pv_file_name>.yaml <1>
```

Please also go through all the values in the PR to ensure they have underscores

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+----
+<1> PVC is in Bound status.
+
+. Create a deployment on Linux by running the following command with following example YAML file:
```
I think this sentence is missing an article here:
```suggestion
. Create a deployment on Linux by running the following command with the following example YAML file:
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+    type: RollingUpdate
+----
+<1> The name of the deployment.
+<2> The volume mount path.
```
Nit - since this callout describes a value that's both on its own and in the middle of a sort of command, perhaps you can reword to add "specify"? That way this label avoids potentially describing that long command as the mount path:
```suggestion
<2> Specify the volume mount path.
```

Up to you though

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-pv.adoc:
```diff
@@ -0,0 +1,192 @@
+...
+----
++
+In this example, there is a _/mnt/smb_ directory mounted as a Common Internet File System (CIFS) filesystem.
```
Since this is a filepath, do you think this should be in backticks instead? I think the [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#quick-markup-reference) count paths as stuff that should have backticks
```suggestion
In this example, there is a `/mnt/smb` directory mounted as a Common Internet File System (CIFS) filesystem.
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,80 @@
+.Prerequisites
+* You are logged in to the running {product-title} cluster.
+
+* You have SMB server installed and know its:
```
Per the [ISG](https://www.ibm.com/docs/en/ibm-style?topic=format-lists#lead-in-wording) we should use complete sentences to introduce lists: 
```suggestion
* You have SMB server installed and know the following information about the server:
```

Feel free to change up some of the wording if you think of a better way to phrase it.

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,80 @@
+* You are logged in to the running {product-title} cluster.
+
+* You have SMB server installed and know its:
+** hostname
+** share name
+** username/password
```
Unless these are specific terms that need to be lowercase, I think these should be capitalized:
```suggestion
** Hostname
** Share name
** Username/password
```

skopacz1 replied:
Suggestion to use "and" instead:
```suggestion
** username and password
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-create-sc.adoc:
```diff
@@ -0,0 +1,80 @@
+** username/password
+
+.Procedure
+To create a storage class:
```
I don't think we typically introduce procedures like this after the .Procedure title, plus I think it's already clear that this is what a user is going to be doing imo. I would suggest removing this line:
```suggestion
```

skopacz1 commented on modules/persistent-storage-csi-smb-cifs-limits.adoc:
```diff
@@ -0,0 +1,15 @@
+
+* FIPS mode is not supported: 
++
+When Federal Information Processing Standards (FIPS) mode is enabled, the use of md4 and md5 are disabled, which prevents users from using ntlm, ntlmv2, or ntlmssp authentication. Also, signing cannot be used since it uses md5. Any CIFS mount that uses these methods fails when FIPS mode is enabled.
```
Per [ISG word usage](https://www.ibm.com/docs/en/ibm-style?topic=word-usage#s), avoid using "since" in place of "because" for translatability purposes:
```suggestion
When Federal Information Processing Standards (FIPS) mode is enabled, the use of md4 and md5 are disabled, which prevents users from using ntlm, ntlmv2, or ntlmssp authentication. Also, signing cannot be used because it uses md5. Any CIFS mount that uses these methods fails when FIPS mode is enabled.
```

skopacz1 commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,39 @@
+
+Familiarity with xref:../../storage/understanding-persistent-storage.adoc#understanding-persistent-storage[persistent storage] and xref:../../storage/container_storage_interface/persistent-storage-csi.adoc#persistent-storage-csi[configuring CSI volumes] is recommended when working with a CSI Operator and driver.
+
+After installing the CIFS/SMB CSI Driver Operator, {product-title} installs corresponding pods for the operator and driver in the `openshift-cluster-csi-drivers` namespace by default. This allows the CIFS/SMB CSI Driver to create CSI-provisioned persistent volumes (PVs) that mount to CIFS/SMB shares.
```
Operator should be capitalized here:
```suggestion
After installing the CIFS/SMB CSI Driver Operator, {product-title} installs corresponding pods for the Operator and driver in the `openshift-cluster-csi-drivers` namespace by default. This allows the CIFS/SMB CSI Driver to create CSI-provisioned persistent volumes (PVs) that mount to CIFS/SMB shares.
```

skopacz1 commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,39 @@
+
+After installing the CIFS/SMB CSI Driver Operator, {product-title} installs corresponding pods for the operator and driver in the `openshift-cluster-csi-drivers` namespace by default. This allows the CIFS/SMB CSI Driver to create CSI-provisioned persistent volumes (PVs) that mount to CIFS/SMB shares.
+
+* The _CIFS/SMB CSI Driver Operator_, after being installed, does not create a storage class by default to use to create persistent volume claims (PVCs). However, xref:../../storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc#persistent-storage-csi-smb-cifs-create-sc_persistent-storage-csi-smb-cifs[you can manually create the CIFS/SMB `StorageClass`]. The CIFS/SMB CSI Driver  ...
```
Just curious, is there a reason that the Operator name is italicized here but not elsewhere?

skopacz1 commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,39 @@
+* The _CIFS/SMB CSI Driver Operator_, after being installed, does not create a storage class by default to use to create persistent volume claims (PVCs). However, xref:../../storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc#persistent-storage-csi-smb-cifs-create-sc_persistent-storage-csi-smb-cifs[you can manually create the CIFS/SMB `StorageClass`]. The CIFS/SMB CSI Driver  ...
+This eliminates the need for cluster administrators to pre-provision storage.
+
+* The _CIFS/SMB CSI driver_ enables you to create and mount CIFS/SMB PVs.
```
Same question as above about italicization, but I'm guessing it's to introduce and differentiate terms

skopacz1 commented on storage/container_storage_interface/persistent-storage-csi-smb-cifs.adoc:
```diff
@@ -0,0 +1,39 @@
+
+include::modules/persistent-storage-csi-smb-cifs-create-pv.adoc[leveloffset=+1]
+
+[role="_additional-resources"]
+== Additional resources
```
Per the [docs guidelines](https://github.com/openshift/openshift-docs/blob/main/contributing_to_docs/doc_guidelines.adoc#anchoring-prerequisites-additional-resources-and-next-steps-titles-in-assemblies), assembly level additional resources section need an anchor ID:
```suggestion
[role="_additional-resources"]
[id="additional-resources_{context}"]
== Additional resources
```

skopacz1 reviewed (commented):
Left some comments but otherwise LGTM!

mpatlasov commented:
LGTM

Phaow commented:
LGTM, thanks for the update!

lpettyjo commented:
@gcharot @mpatlasov @Phaow New Static and Dynamic Provisioning sections added for you to review. However, I notice now that the PVC doesn't go into Bound state when following procedure in Static section. Only change here was removing use of SC.

gcharot commented on modules/persistent-storage-csi-smb-cifs-provision-dynamic.adoc:
```diff
@@ -0,0 +1,130 @@
+
+. Create a PVC:
+
+.. Create a PVC by running the following command with the following example YAML file:
```
We need to highlight somewhere that provisioning volumes will will create a sub directory with the PV name under the "source" defined in the SC

lpettyjo replied:
@gcharot This is mentioned in callout 3 under the SC YAML file in step 2. If this is not visible enough, I can pull it out into the text below the command for this step (this will make it more visible) and/or move it to step 3 when creating the PVC, if you think that would be helpful.

gcharot replied:
Correct I missed it. We can keep it there but also add it to the Dynamic provisioning intro sentence?

lpettyjo replied:
Done.

gcharot commented on modules/persistent-storage-csi-smb-cifs-provision-static.adoc:
```diff
@@ -0,0 +1,187 @@
+      namespace: <namespace> <5>
+----
+<1> The name of the PV.
+<2> `volumeHandle` format: {smb-server-address}#{sub-dir-name}#{share-name}. Ensure that this value is unique for every share in the cluster.
```
Why do we have a "#" sign instead of a "."?

lpettyjo replied:
Weirdly, that is what is called out in the K8s example. However, the actual YAML file content uses dot, and that is what I used when testing it, so I'll change it.

gcharot commented on modules/persistent-storage-csi-smb-cifs-provision-static.adoc:
```diff
@@ -0,0 +1,187 @@
+
+To set up static provisioning:
+
+. Create a PV by running the following command with the following example YAML file:
```
Do we want to include the secret creation again or do we assume it's clear from the last step? Knowing that it's not necessary the same secret as dynamic provisioning.

lpettyjo replied:
I think that secret creation should be in the procedure, so I've added it.

lpettyjo commented on modules/persistent-storage-csi-smb-cifs-provision-static.adoc:
```diff
@@ -0,0 +1,187 @@
+.Prerequisites
+* Access to the {product-title} web console.
+* {FeatureName} CSI Driver Operator and driver installed.
+
```
Done.

lpettyjo commented on modules/persistent-storage-csi-smb-cifs-provision-static.adoc:
```diff
@@ -0,0 +1,187 @@
+  annotations:
+    pv.kubernetes.io/provisioned-by: smb.csi.k8s.io
+  name: <pv_name> <1>
+spec:
```
Done.

gcharot commented on modules/persistent-storage-csi-smb-cifs-provision-static.adoc:
```diff
@@ -0,0 +1,216 @@
+  accessModes:
+    - ReadWriteMany
+  persistentVolumeReclaimPolicy: Retain
+  storageClassName: ""
```
@mpatlasov Can you please make sure this actually works?

cc @Phaow

Phaow replied:
Hi @gcharot @mpatlasov , I could confirm that if we use the `storageClassName: ""` in the `PersistentVolume` then we also need set the same storageclass name in the pvc definition. Otherwise `persistentvolume-controller` will complain `Cannot bind to requested volume "manual-pv": storageClassName does not match`.

gcharot replied:
thanks @Phaow good call!

lpettyjo replied:
Seems to work for me too. PVC goes to Bound state.

Phaow commented on modules/persistent-storage-csi-smb-cifs-provision-static.adoc:
```diff
@@ -0,0 +1,216 @@
+  resources:
+    requests:
+      storage: <storage_amount> <2>
+  volumeName: <pv_name> <3>
```
```suggestion
  storageClassName: ""
  volumeName: <pv_name> <3>
```
We also need add the storageClassName at here.

lpettyjo replied:
Done.

Phaow commented on modules/persistent-storage-csi-smb-cifs-provision-static.adoc:
```diff
@@ -0,0 +1,216 @@
+----
+Name:          pvc-test
+Namespace:     default
+StorageClass:  smb
```
```suggestion
StorageClass:  
```
Since we used the `storageClassName: ""`, keep consistent will be better.

lpettyjo replied:
Done.

lpettyjo commented on modules/persistent-storage-csi-smb-cifs-provision-dynamic.adoc:
```diff
@@ -0,0 +1,130 @@
+----
+Name:          pvc-test
+Namespace:     default
+StorageClass:  smb
```
Actually, that file is now deleted, and the SC example YAML file doesn't specify a name, but I changed this anyhow.

lpettyjo commented:
/retest

Phaow commented:
Others LGTM, thanks for the update again!

lpettyjo commented on modules/persistent-storage-csi-smb-cifs-limits.adoc:
```diff
@@ -0,0 +1,15 @@
++
+When Federal Information Processing Standards (FIPS) mode is enabled, the use of md4 and md5 are disabled, which prevents users from using ntlm, ntlmv2, or ntlmssp authentication. Also, signing cannot be used because it uses md5. Any CIFS mount that uses these methods fails when FIPS mode is enabled.
+
+* HTTP proxy configuration is not supported.
```
Is this OK?

* Using HTTP proxy configuration to connect to outside of the cluster SMB servers is not supported by the CSI driver.
+
Since CIFS/SMB is a LAN protocol, and though it can be routed to subnets, it is not designed to be extended over the WAN, and does not support HTTP proxy settings.

Phaow replied:
Looks good.

mpatlasov commented:
I added a comment above to add text clarifying that the deployment in static provisioning section is just an example, not something mandatory. Other than that, LGTM! (yaml files look fine now)

gcharot commented:
LGTM

Phaow commented:
LGTM

lpettyjo commented:
> I added a comment above to add text clarifying that the deployment in static provisioning section is just an example, not something mandatory. Other than that, LGTM! (yaml files look fine now)

@mpatlasov I can't find your comment, so I added a note essentially quoting what you said in this statement. Let me know if this is OK.

lpettyjo commented:
/cherrypick enterprise-4.16
